Update cfitsio to 3.41

mhvk opened the pull request:
This updates `cfitsio` to v3.41. One immediate question: I added a script that separates out the library files actually needed from sample programmes. Should we still include those additional files (also `README`, etc.)? I can easily not include them by removing the third commit, with possible adjustment of `trim_cfitsio.sh` to do this automatically.

mhvk commented:
Hmm, appveyor fails with
```
<stdout>(720) : fatal error C1083: Cannot open include file: 'unistd.h': No such file or directory
```
This file is not shipped with `cfitsio`

Hmm, seems it should have been excluded on windows (in `zconf.h`)

mhvk commented:
cc @astrofrog, @eteq for the what-to-include packaging question

Anybody who can help with the windows issue (linux seems fine; still waiting on MacOS)

MSeifert04 commented on cextern/cfitsio/CMakeLists.txt:
```diff
@@ -0,0 +1,112 @@
+    # Define
+    ADD_DEFINITIONS(-D_CRT_SECURE_NO_DEPRECATE)
+    # Need an empty unistd.h in MSVC for flex-generated eval_l.c:
+    FILE(WRITE ${CMAKE_SOURCE_DIR}/unistd.h "")
```
Shouldn't take care of the missing `unistd.h` file?

mhvk replied:
None of the `Makefile` get executed/included (certainly not now, I don't think before either - `setup_package` just includes `*.c` files), but obviously it should not be hard to mimic this in `setup_package`. I don't really understand, though, why this worked before...

MSeifert04 commented:
> Should we still include those additional files (also README, etc.)?

Depends a bit on the built time. They probably won't hurt but if they significantly increase the build-time I'd say we should remove them again.

mhvk commented:
The other files do not get built at all - everything relevant is moved to a new `lib` directory. I mostly wondered what best practice was - at some level one wants to stay as close to the original package as possible.

astrofrog commented:
I don't feel strongly about this - @eteq @taldcroft?

mhvk commented:
OK, now got the windows build working as well. I'm Cc'ing @embray as he's probably the last one to have done this and might have comments. Overall, it seems OK, but suggestions are still very welcome on the setup with `setup_package.py` and `trim_cfitsio.sh`. Also, are the comments in those files clear enough for future updates?

saimn commented:
> The other files do not get built at all - everything relevant is moved to a new lib directory. I mostly wondered what best practice was - at some level one wants to stay as close to the original package as possible.

I have no strong preference, on one hand staying close to the original package makes it easier to update, on the other hand we need only a subset of these files and some tweaks are needed anyway to compile cfitsio when included in Astropy (like moving files in a `lib/` subdirectory). I think I would go for cleaning unused files (automake/cmake files, docs, scripts etc.).

MSeifert04 commented:
Okay, let's stick with the required files and remove all the (new) unused ones. Is that easily possible @mhvk?

mhvk commented:
@MSeifert04 - yes, not hard to remove all files, since I kept that as a single commit. Now removed them all, and adjusted the `trim_cfitsio` script to do that automatically. Hopefully, all tests will still pass...

saimn commented:
Thanks @mhvk , tests pass so I guess it's good to go :sweat_smile: 
Ok for you @MSeifert04 ?

MSeifert04 commented on cextern/trim_cfitsio.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/sh
+
+# This script should be run every time cfitsio is updated.
+# This moves allthe code needed for the actual library to lib
```
missing whitespace between all and the.

mhvk commented:
@MSeifert04 - OK, done as part of the rebase...

MSeifert04 commented:
Thanks @mhvk !
